Src/libCZI/utilities.cpp

```diff
@@ -423,3 +424,158 @@ tString trimImpl(const tString& str, const tString& whitespace)
+        if (rectangle_b.y < rectangle_a.y)
+        {
+            result[result_index++] = libCZI::IntRect{ rectangle_a.x, rectangle_b.y, rectangle_b.x + rectangle_b.w - rectangle_a.x, rectangle_a.y - rectangle_b.y };
+        }
+
+        if (rectangle_b.y + rectangle_b.h > rectangle_a.y + rectangle_a.h)
+        {
+            result[result_index++] = libCZI::IntRect{ rectangle_a.x, rectangle_a.y + rectangle_a.h, rectangle_b.x + rectangle_b.w - rectangle_a.x, rectangle_b.y + rectangle_b.h - rectangle_a.y - rectangle_a.h };
+        }
+    }
+    else if (rectangle_b.x >= rectangle_a.x && rectangle_b.x + rectangle_b.w > rectangle_a.x + rectangle_a.w)
+    {
+        result[result_index++] = libCZI::IntRect{ rectangle_a.x + rectangle_a.w, rectangle_b.y, rectangle_b.x + rectangle_b.w - rectangle_a.x - rectangle_a.w, rectangle_b.h };
+
+        if (rectangle_b.y < rectangle_a.y)
+        {
+            result[result_index++] = libCZI::IntRect{ rectangle_b.x, rectangle_b.y, rectangle_a.x + rectangle_a.w - rectangle_b.x, rectangle_a.y - rectangle_b.y };
+        }
+
+        if (rectangle_b.y + rectangle_b.h > rectangle_a.y + rectangle_a.h)
+        {
+            result[result_index++] = libCZI::IntRect{ rectangle_b.x, rectangle_a.y + rectangle_a.h, rectangle_a.x + rectangle_a.w - rectangle_b.x, rectangle_b.y + rectangle_b.h - rectangle_a.y - rectangle_a.h };
+        }
+    }
+    else if (rectangle_b.x <= rectangle_a.x && rectangle_b.x + rectangle_b.w >= rectangle_a.x + rectangle_a.w)
+    {
+        result[result_index++] = libCZI::IntRect{ rectangle_b.x, rectangle_b.y, rectangle_a.x - rectangle_b.x, rectangle_b.h };
+        result[result_index++] = libCZI::IntRect{ rectangle_a.x + rectangle_a.w, rectangle_b.y, rectangle_b.x + rectangle_b.w - rectangle_a.x - rectangle_a.w, rectangle_b.h };
+
+        if (rectangle_a.y > rectangle_b.y)
+        {
+            result[result_index++] = libCZI::IntRect{ rectangle_a.x, rectangle_b.y, rectangle_a.w, rectangle_a.y - rectangle_b.y };
+        }
+        else if (rectangle_a.y + rectangle_a.h > rectangle_b.y && rectangle_a.y + rectangle_a.h < rectangle_b.y + rectangle_b.h)
+        {
+            result[result_index++] = libCZI::IntRect{ rectangle_a.x, rectangle_a.y + rectangle_a.h, rectangle_a.w, rectangle_b.y + rectangle_b.h - rectangle_a.y - rectangle_a.h };
+        }
+    }
+
+    return result_index;
+}
+
+
+/*static*/bool RectangleCoverageCalculator::IsCompletelyContained(const libCZI::IntRect& outer, const libCZI::IntRect& inner)
+{
+    return inner.x >= outer.x && inner.x + inner.w <= outer.x + outer.w &&
+        inner.y >= outer.y && inner.y + inner.h <= outer.y + outer.h;
+}
+
+std::int64_t RectangleCoverageCalculator::CalcAreaOfIntersectionWithRectangle(const libCZI::IntRect& query_rectangle) const
+{
+    if (!query_rectangle.IsValid())
+    {
+        return 0;
```

Comment:
Do we have a use-case for passing invalid rectangles to this function or ought we outright throw an error?

Reply:
Just posing the question here even though the convention for this library has been to have a 0 return.

Reply:
Agreed, this is a bit a misnomer. The IntRect-struct considers a rectangle with width or height <= 0 [as invalid](https://github.com/ZEISS/libczi/blob/main/Src/libCZI/libCZI_Pixels.h). We'd better distinguish between "empty" and "invalid", and I'd guess one would not consider an empty rectangle as invalid. So, in this case, I'd think: passing in an empty rectangle is ok (and we should deal with this), but with an invalid rectangle we might want to bail out.